### ISSUE TITLE
Causal Convolution

### ISSUE THREAD

**Issue by calclavia** [pytorch/pytorch#1333]
Is there currently a way to specify causal convolutions in PyTorch?

Keras has a way to specify a "causal" padding in a conv layer https://keras.io/layers/convolutional/#conv1d

Note: I'm trying to implement Wavenet (https://deepmind.com/blog/wavenet-generative-model-raw-audio/).

**Comment by apaszke** [comment]
Isn't it enough to use `F.pad` and prepend the `dilation` amount to the sequence?

**Comment by calclavia** [comment]
@apaszke 
Hmm, I think you're right. A kernel=2 conv with padding should achieve the same effect. Problem is, I can't use `F.pad` with 1D time series data.

I think this code should be a workaround:

dilation = 1:
```
x = torch.autograd.Variable(torch.FloatTensor([[[1, 2, 3, 4]]]))
x = torch.cat((torch.autograd.Variable(torch.zeros(1, 1, 1)), x), 2)
c = nn.Conv1d(1, 1, 2)
c(x)
```

Is there a more elegant way to do it?

**Comment by calclavia** [comment]
Alright, I think I found a better way:

```
x = torch.autograd.Variable(torch.FloatTensor([[[1, 2, 3, 4]]]))
c = nn.Conv1d(1, 1, 2, padding=1)
out = c(x)[:, :, :-1]
```

**Comment by neverfox** [comment]
Or you could generalize:

```
class CausalConv1d(nn.Conv1d):
    def __init__(self,
                 in_channels,
                 out_channels,
                 kernel_size,
                 stride=1,
                 dilation=1,
                 groups=1,
                 bias=True):
        super(CausalConv1d, self).__init__(in_channels, out_channels,
                                           kernel_size, stride, 0, dilation,
                                           groups, bias)

        self.left_padding = dilation * (kernel_size - 1)

    def foward(self, input):
        n, c, _ = input.size()
        input = F.pad(input.unsqueeze(2),
                      (self.left_padding, 0, 0, 0)).view(n, c, -1)

        return super(CausalConv1d, self).call(input)
```

**Comment by Maltimore** [comment]
I think it's not correct that you can get causal convolutions by way of padding. At the 'end' of the input sequence this works, but not for intermediate elements of the convolution. These will 'see' the future.
@apaszke

**Comment by arogozhnikov** [comment]
@apaszke does this introduce an additional copy (compared to just conv1d?)

**Comment by arogozhnikov** [comment]
As a proposed workaround - pad from both sides in convolution, then take a slice on output.

**Comment by soheilzabihi** [comment]
Dear @arogozhnikov,
In this implementation we have aditional computation, Moreover, it also introduces an additional copy, you can check it in the following:
`class CausalConv1d(torch.nn.Conv1d):
    def __init__(self, in_channels, out_channels, kernel_size,
                 stride=1, dilation=1, groups=1, bias=True):
        self.__padding = (kernel_size - 1) * dilation
        super(CausalConv1d, self).__init__(in_channels, out_channels, kernel_size=kernel_size,
                                           stride=stride, padding=self.__padding, dilation=dilation,
                                           groups=groups, bias=bias)

    def forward(self, input):
        result = super(CausalConv1d, self).forward(input)
        print(id(result))
        if self.__padding != 0:
            return result[:, :, :-self.__padding]
        return result`

`CaConv1d = CausalConv1d(in_channels=2, out_channels=6, kernel_size=2, dilation=1)
out = CaConv1d(t4d)
print(id(out))`
output:
`140510012165504
140510011010144`

as you can see, the objects references are different. 
but when we do not have padding it is correct.
`CaConv1d = CausalConv1d(in_channels=2, out_channels=6, kernel_size=1, dilation=1)
out = CaConv1d(t4d)
print(id(out))` 
output:
`140510011660256
140510011660256`
so, these are same python object.

**Comment by soheilzabihi** [comment]
Dear @arogozhnikov,
In this implementation, we have additional computation because of right side padding of the input, which dosen't require in the Casual Convolution. Moreover, when `self.__padding` is not zero it also introduces an additional copy in compared to conv1d. you can check it in the following 
- just `print(id(result))` is added to the `forward` method of the @arogozhnikov's implimentation
```python
class CausalConv1d(torch.nn.Conv1d):
    def __init__(self,
                 in_channels,
                 out_channels,
                 kernel_size,
                 stride=1,
                 dilation=1,
                 groups=1,
                 bias=True):
        self.__padding = (kernel_size - 1) * dilation

        super(CausalConv1d, self).__init__(
            in_channels,
            out_channels,
            kernel_size=kernel_size,
            stride=stride,
            padding=self.__padding,
            dilation=dilation,
            groups=groups,
            bias=bias)

    def forward(self, input):
        result = super(CausalConv1d, self).forward(input)
        print(id(result))  # printing object refrence
        if self.__padding != 0:
            return result[:, :, :-self.__padding]
        return result
```
```python
input = torch.from_numpy(np.ones((3,2,5))).float()
CaConv1d = CausalConv1d(in_channels=2, out_channels=6, kernel_size=2, dilation=1)
out = CaConv1d(input)
print(id(out))
```
output:
```python
140510012165504
140510011010144
```
as you see, the objects references are different; thetefore, new object is created.
but when we do not have padding it is correct.
```python
CaConv1d = CausalConv1d(in_channels=2, out_channels=6, kernel_size=1, dilation=1) out = CaConv1d(input) 
print(id(out))
```
output:
```python
140510011660256
140510011660256
```
so, these are same python object.

**Comment by soheilzabihi** [comment]
By changing @neverfox in the following way, we can avoid wasting memory. It's not worse to say that there is not additional computation because just left side is padded.
```python
class CausalConv1d(torch.nn.Conv1d):
    def __init__(self,
                 in_channels,
                 out_channels,
                 kernel_size,
                 stride=1,
                 dilation=1,
                 groups=1,
                 bias=True):

        super(CausalConv1d, self).__init__(
            in_channels,
            out_channels,
            kernel_size=kernel_size,
            stride=stride,
            padding=0,
            dilation=dilation,
            groups=groups,
            bias=bias)
        
        self.__padding = (kernel_size - 1) * dilation
        
    def forward(self, input):
        return super(CausalConv1d, self).forward(F.pad(input, (self.__padding, 0)))
```

**Comment by yuhonghong66** [comment]
This would be fantastic to have structure so that all edits are forced to be spreads...

**Comment by arogozhnikov** [comment]
Hi @soheilzabihi ,
 
Thanks for checking! There is a problem however - you check ids - but id isn't connected to actual memory management. 
In example below x and y share same memory (check it!), but their `id` are different.

```python
In [2]: import torch
In [3]: x = torch.zeros([10000])
In [4]: y = x[:]
In [5]: id(y) - id(x)
```
So this example we can see that slicing (e.g. `result[:, :, :-self.__padding]`) does not waste memory.

And `F.pad` is actually where you get new memory allocated. Though it may be freed later (I didn't check) - but you will definitely have to allocate it at least.

**Comment by soheilzabihi** [comment]
Hi @arogozhnikov ,
You are right, `id` doesn't show the physical memory address.
But, by using `result[:, :, :-self.__padding]` to drop the right side after 1d convolution, we have excess computation, although it's not a big deal when GPU is used.

**Comment by rebeen** [comment]
Sorry could you let me know which one could be used properly without increasing the computation time compared to the actual con1d 

thank you

**Comment by PraljakReps** [comment]
> Hi @SoheilZabihi ,
> 
> Thanks for checking! There is a problem however - you check ids - but id isn't connected to actual memory management. In example below x and y share same memory (check it!), but their `id` are different.
> 
> ```python
> In [2]: import torch
> In [3]: x = torch.zeros([10000])
> In [4]: y = x[:]
> In [5]: id(y) - id(x)
> >>> 1600427632   # output varies, but its not zero
> ```
> 
> So this example we can see that slicing (e.g. `result[:, :, :-self.__padding]`) does not waste memory.
> 
> And `F.pad` is actually where you get new memory allocated. Though it may be freed later (I didn't check) - but you will definitely have to allocate it at least.

I just want to emphasize this succinct analysis; F.pad() can lead to memory allocation or even non-deterministic operations during the backboard pass, leading to OOM for large models or a lack of reproducible code. I experienced this issue when using a conditional-Wavenet model. For the latter case (non-deterministic behavior), this happens specifically on Cuda backend and during backdrop -- check the documentation: https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html
